GVR compare on plan reconcile

Danil-Grigorev opened the pull request:
Allows to compare supported GVRs between clusters, and provides per-namespace report on resources, which will need APIGroup conversion or manual exclusion from the current migration run. Those resources will not be restored on the destination cluster without manual conversion, so the migration will not be fully successful.

Sample output:
```yaml
  observedGeneration: 16674
  unsupportedNamespaces:
  - name: phronetic
    unsupportedResources:
    - batch/v2alpha1, Resource=cronjobs
    - batch/v2alpha1, Resource=scheduledjobs
```

Output from OCP4.4 (kube 16+)
```yaml
    incompatibleNamespaces:
  - gvrs:
    - group: extensions
      resoruce: daemonsets
      version: v1beta1
    - group: extensions
      resoruce: deployments
      version: v1beta1
    - group: extensions
      resoruce: replicasets
      version: v1beta1
    - group: apps
      resoruce: controllerrevisions
      version: v1beta1
    - group: apps
      resoruce: deployments
      version: v1beta1
    name: openshift-migration
  - gvrs:
    - group: extensions
      resoruce: deployments
      version: v1beta1
    - group: extensions
      resoruce: replicasets
      version: v1beta1
    - group: apps
      resoruce: deployments
      version: v1beta1
    name: nginx-deployment
  - gvrs:
    - group: batch
      resoruce: cronjobs
      version: v2alpha1
    - group: batch
      resoruce: scheduledjobs
      version: v2alpha1
    name: cron

```

Warning message:
```yaml
  - category: Warn
    lastTransitionTime: 2020-02-17T12:55:52Z
    message: 'Some namespaces are unsupported. See: `unsupportedNamespaces` for
      details.'
    status: "True"
    type: NotSupported
```

Demo: https://drive.google.com/open?id=1J7_pA5sRSpnuR17JiGiLVaVfT8iNIPZW

Danil-Grigorev commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+	unsupportedGVRs, err := r.unsupportedServerResources(gvDiff)
+	if err != nil {
+		return errors.Wrap(err, "Unable to get unsupported resources for scrCluster")
+	}
```
Both methods will reduce the number of resources, which will be listed in each `migPlan` namespace to the minimum amount, allowing to inspect every namespace in details by listing for unsupported resources in it and checking for their presence.

Danil-Grigorev commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+	return append(missingGroups, missingVersions...), nil
+}
+
+func (r *GVKCompare) collectNamespaceReport(unsupportedResources []schema.GroupVersionResource) error {
```
Collects report for `MigPlan` namespaces only for unsupported resources.

Danil-Grigorev commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+		unsupportedGVRs := []string{}
+		for _, gvr := range unsupportedResources {
+			options := metav1.ListOptions{}
+			resourceList, err := r.SrcClient.Resource(gvr).Namespace(namespace).List(options)
```
Dynamic client was the simplest option to allow generic listing of unsupported resources on the source cluster, as it does not require typecasting, and encapsulates constructing of `*Unstructured` object.

Danil-Grigorev commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+			}
+
+			r.APIResources = namespaced(r.APIResources)
+			r.APIResources = excludeSubresources(r.APIResources)
```
Removes unnecessary [`resource/status`, `resource/logs`...] api paths.

Danil-Grigorev commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+	})
+	sort.Slice(reducedDst, func(i int, j int) bool {
+		return reducedDst[i].Name < reducedDst[j].Name
+	})
```
Ensure the order and number of `APIGroups` for both cluster.

djwhatle commented on config/crds/migration_v1alpha1_migcluster.yaml:
```diff
@@ -58,6 +58,12 @@ spec:
           type: object
         status:
+          properties:
+            observedGeneration:
```
This and other CRD changes are just a leftover from artifacts not being checked in on https://github.com/fusor/mig-controller/pull/412/files

Danil-Grigorev replied:
Don't mind importing them with the PR, as it still requires a CRD change.

jortel replied:
Ah, looks like I forget the `make manfests` on pull/412.  Let's fix this in a separate PR and then rebase.  The reason for not combining is that if this PR got reverted, we would not want the CRD change to be reverted.

jortel replied:
Fixed on master: https://github.com/fusor/mig-controller/pull/421

djwhatle commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+		plan.Status.SetCondition(migapi.Condition{
+			Type:     NotSupported,
+			Status:   True,
+			Category: Warn,
```
Note to anyone else reviewing, this condition is a `Warn` at present. 
cc @jortel

jortel commented on pkg/apis/migration/v1alpha1/gvk.go:
```diff
@@ -0,0 +1,29 @@
+// to contain resources unsupported by the migration
+type UnsupportedNamespace struct {
+	Name                 string   `json:"name"`
+	UnsupportedResources []string `json:"unsupportedResources"`
```
This should just be `Resources` within the scope of `UnsupportedNamespaces`.

Danil-Grigorev replied:
This will complicate understanding for this field in the status. Either the name, or some sub item should tell the reason for it being there.

jortel commented on pkg/apis/migration/v1alpha1/gvk.go:
```diff
@@ -0,0 +1,29 @@
+	return namespaces
+}
+
+func (p *MigPlanStatus) InitUnsupported() {
```
:+1: for functional decomposition (generally) but don't think this function adds value.

Danil-Grigorev replied:
I consider this an initial run for GVK related tasks. This structure may contain additional fields, some specific info to `why` the resource is unsupported, and what needs to be done, which it does not have at the moment. May be the name change to `ResetUnsupported()` will be sufficient?

jortel commented on pkg/apis/migration/v1alpha1/gvk.go:
```diff
@@ -0,0 +1,29 @@
+	p.UnsupportedNamespaces = []UnsupportedNamespace{}
+}
+
+func (p *MigPlanStatus) AppendUnsupported(namespace string, resources []string) {
```
Perhaps add:
```
if p.UnsupportedNamespaces == nil {
  p.UnsupportedNamespaces = []UnsupportedNamespace{}
}
```
to avoid the need for 2 step construction.

jortel replied:
Rather than appending the slice in `MigPlanStatus` you can eliminate some complexity and coupling by building a []UnsupportedNamespace{} then assigning to `MigPlanStatus` already populated.

jortel commented on pkg/cloudprovider/aws.go:
```diff
@@ -187,16 +187,16 @@ func (p *AWSProvider) UpdateRegistryDC(dc *appsv1.DeploymentConfig, name, dirNam
 		dc.Spec.Template.Spec.Containers[0].VolumeMounts = append(
 			dc.Spec.Template.Spec.Containers[0].VolumeMounts,
 			kapi.VolumeMount{
```
Let's fix this unrelated formatting in a separate PR and rebase.

jortel commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+
+	gvkCompare, err := r.prepareGVKCompare(plan)
+	if err != nil {
+		log.Error(err, "Failed to prepare GVK compare")
```
The convention in the controller is to `log.Trace(err)` instead.  The trace provides all the necessary information (and more).

jortel commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+
+	err = gvkCompare.Compare()
+	if err != nil {
+		log.Error(err, "Failed to compare GVRs on both clusters")
```
The convention in the controller is to `log.Trace(err)` instead.  The trace provides all the necessary information (and more).

jortel commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	err := gvkCompare.PrepareSourceDiscovery(r)
+	if err != nil {
```
Please add:
```
log.Trace(err)
```

Danil-Grigorev replied:
Should pick it up: https://github.com/fusor/mig-controller/pull/413/files#diff-ff4cf8d4f8669d3270d95d4c88c43266R17

jortel commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	err = gvkCompare.PrepareDestinationDiscovery(r)
+	if err != nil {
```
Please add:
```
log.Trace(err)
```

jortel commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	err = gvkCompare.PrepareSourceClient(r)
+	if err != nil {
```
Please add:
```
log.Trace(err)
```

jortel commented on pkg/controller/migplan/migplan_controller.go:
```diff
@@ -254,6 +254,12 @@ func (r *ReconcileMigPlan) Reconcile(request reconcile.Request) (reconcile.Resul
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
```
To be consistent with other validations and to keep `Reconcile()` clean, let's move this into:  `ReconcileMigPlan.validate()`.

jortel commented on pkg/controller/migplan/validation.go:
```diff
@@ -94,6 +95,7 @@ const (
 	StorageNotReadyMessage                = "The referenced `migStorageRef` does not have a `Ready` condition."
 	NsListEmptyMessage                    = "The `namespaces` list may not be empty."
 	InvalidDestinationClusterMessage      = "The `srcMigClusterRef` and `dstMigClusterRef` cannot be the same."
+	NsNotSupported                        = "Namespaces [] not supported. See: `unsupportedNamespaces` for details."
```
Since there is a separate list (which may be large) in the status, let's not list them on the condition as well.

jortel commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+	return nil
+}
+
+func (r *GVKCompare) PrepareSourceDiscovery(c client.Client) error {
```
Instead of Prepare__() methods, I think the idiomatic way to accomplish complex construction would be to use a `NewComparator() (*Comparator, error)` function.

jortel commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+package gvkcompare
```
Perhaps the package named `gvk` would better support a broader range of tools in the future.
Within the package, suggest a `comparator.go` file would be intuitive.

jortel replied:
Actually, another question.  If only used by the plan controller, why is this a top level package instead of a sub-package under the plan controller package?

jortel commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+)
+
+// GVKCompare reconciles a GVKCompare object
+type GVKCompare struct {
```
Within the `gvk` package, suggest the type named `Comparator` would be intuitive.  Would be used as: `gvk.Comparator{}`

jortel commented on pkg/gvkcompare/gvkcompare.go:
```diff
@@ -0,0 +1,276 @@
+}
+
+func (r *GVKCompare) collectNamespaceReport(unsupportedResources []schema.GroupVersionResource) error {
+	for _, namespace := range r.Plan.GetSourceNamespaces() {
```
This appears to make an API call for: (each namespace on the plan) **x** (all GVK combinations identified as missing on the destination).  Right?
It may be more efficient to query the GVK across all namespaces and filter/collate on namespace.

Danil-Grigorev replied:
This appeared to me as an option. It sounds better then per-namespace requests, will update as the initial approach is now functional.

djwhatle commented:
@Danil-Grigorev are vendor updates needed for this PR? I'm unable to build it locally.

```
        imports github.com/evanphx/json-patch: cannot find package "github.com/evanphx/json-patch" in any of:
        /home/dwhatley/dev/GO/src/github.com/fusor/mig-controller/vendor/github.com/evanphx/json-patch (vendor tree)
        /usr/local/go/src/github.com/evanphx/json-patch (from $GOROOT)
        /home/dwhatley/dev/GO/src/github.com/evanphx/json-patch (from $GOPATH)
```

djwhatle commented:
@Danil-Grigorev I'm having trouble getting a successful connection between clusters after I switched to this build of mig-controller. Any special instructions for testing? I installed using non-olm 1.1 manifests on both a 3.7 / 4.2 combo, setting these values on the `MigrationController` CR on both ends.

```
  mig_controller_image: quay.io/dgrigore/mig-controller
  mig_controller_version: latest
```

djwhatle commented:
Going to take another look at this PR today, will see if I can get it to run successfully and check if I see the observedGeneration increasing too quickly.

djwhatle commented:
@Danil-Grigorev I'm getting an error starting up remote watches on a current build of this. I guess the "work-in-progress" label means something here :)

```
{"level":"info","ts":1582755048.009608,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"migplan-controller","worker count":1}
E0226 22:10:48.111303       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1/migcluster_types.go:151
/go/src/github.com/konveyor/mig-controller/pkg/gvk/compare.go:111
/go/src/github.com/konveyor/mig-controller/pkg/gvk/compare.go:66
/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:38
/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:16
/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/validation.go:188
/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:217
/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x108 pc=0x149ce47]

```

djwhatle commented:
@Danil-Grigorev @jortel ran the latest commit of this and had it successfully detect GVK differences across my clusters.

```
    unsupportedNamespaces:
    - name: phronetic
      unsupportedResources:
      - batch/v2alpha1, Resource=cronjobs
      - batch/v2alpha1, Resource=scheduledjobs
```

@Danil-Grigorev did you ever figure out what was up with the inability to detect one of the GVK differences between the 3.x and 4.4 pair we looked at on our review call earlier?

djwhatle commented on pkg/controller/migmigration/quiesce.go:
```diff
@@ -88,25 +102,65 @@ func (t *Task) scaleDownDeploymentConfigs(client k8sclient.Client) error {
+}
+
+// Scales down all Deployments (kubernetes < 16)
+func (t *Task) scaleDownv1betaDeployments(client k8sclient.Client) error {
```
I have the same question for here. If we are going with the approach Jeff outlined that automatically handles older versions, would these version-specialized methods still be necessary?

djwhatle commented on pkg/controller/migmigration/quiesce.go:
```diff
@@ -182,26 +269,66 @@ const (
+}
+
+// Scales down all DaemonSets. (kubernetes < 16)
+func (t *Task) scaleDownv1betaDaemonSets(client k8sclient.Client) error {
```
Same question here.

djwhatle commented on pkg/gvk/compare.go:
```diff
@@ -0,0 +1,280 @@
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+)
+
+var crdGVR = schema.GroupVersionResource{
```
In case anyone else reviewing  is wondering: this is used for excluding all CRDs from the comparison list since CRDs can be moved over to a dest cluster already.

djwhatle commented on pkg/gvk/compare.go:
```diff
@@ -0,0 +1,280 @@
+	gvDiff := compareGroupVersions(srcGroupList, dstGroupList)
+	unsupportedGVRs, err := r.unsupportedResources(gvDiff)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get unsupported resources for scrCluster")
```
typo: `scrCluster` -> `srcCluster`

Danil-Grigorev replied:
:+1:

djwhatle commented on pkg/gvk/compare.go:
```diff
@@ -0,0 +1,280 @@
+	return unsupportedNamespaces, nil
+}
+
+func (r *Compare) occureIn(gvr schema.GroupVersionResource) ([]string, error) {
```
typo: `occureIn`->`occurIn`

Danil-Grigorev replied:
:+1:

djwhatle commented on pkg/gvk/compare.go:
```diff
@@ -0,0 +1,280 @@
+	for _, crd := range crdList.Items {
+		group, found, err := unstructured.NestedString(crd.Object, groupPath...)
+		if !found {
+			return errors.Wrap(err, "Error while extracting CRD group: not exist")
```
"not exist" -> "doesn't exist"

Danil-Grigorev replied:
Thanks

djwhatle commented on pkg/gvk/compare_test.go:
```diff
@@ -0,0 +1,216 @@
+)
+
+var (
+	sourceGroups = []metav1.APIGroup{
```
@Danil-Grigorev big 👍 for adding tests!

Danil-Grigorev replied:
Removing those, as unnecessary addition. The data set for the tests is to huge.

djwhatle commented on pkg/health/pods.go:
```diff
@@ -87,15 +90,34 @@ func PodManagersRecreated(client k8sclient.Client, options *k8sclient.ListOption
-	daemonSetList := v1beta1.DaemonSetList{}
-
-	err := client.List(context.TODO(), options, &daemonSetList)
+func DaemonSetsRecreated(
```
@Danil-Grigorev same question here: is this necessary if we use @jortel change that auto-downgrades versions when necessary?

Danil-Grigorev replied:
It is unrelated to the PR

Danil-Grigorev commented on pkg/apis/migration/v1alpha1/migcluster_types.go:
```diff
@@ -174,23 +188,45 @@ func (m *MigCluster) DeleteResources(client k8sclient.Client, labels map[string]
+	options := k8sclient.MatchingLabels(labels)
+
+	// Deployment
+	if kubeVersion >= reference.AppsGap {
```
It is unrelated to the PR

Danil-Grigorev commented on pkg/controller/migmigration/migrate.go:
```diff
@@ -67,13 +68,13 @@ func (r *ReconcileMigMigration) migrate(migration *migapi.MigMigration) (time.Du
-		Phase:           migration.Status.Phase,
-		Annotations:     r.getAnnotations(migration),
-		BackupResources: r.getBackupResources(migration),
+		ReconcileMigMigration: r,
```
Sorry for confusion @djwhatle, the client-go scheme is created as a single instance on the client-go library import, and all the managers are referencing it and not creating a new one each time. @jortel approach is correct.

Danil-Grigorev commented on pkg/controller/migmigration/migmigration_controller.go:
```diff
@@ -45,12 +45,12 @@ func Add(mgr manager.Manager) error {
 
 // newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+func newReconciler(mgr manager.Manager) *ReconcileMigMigration {
```
There is no reason to do that. Unrelated to the PR

Danil-Grigorev commented on pkg/controller/migmigration/migmigration_controller.go:
```diff
@@ -25,7 +25,7 @@ import (
 	migref "github.com/konveyor/mig-controller/pkg/reference"
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
It's IDE configuration, have no idea.

Danil-Grigorev commented:
@djwhatle Would have to ask for a final review. 

After yesterday's testing I found out that filtering based on GroupVersions provided form the server is not sufficient. The issue you saw was due to the fact that group extensions/v1beta1 never went away in a newer cluster versions, just it's content changed. So my algorithm was marking as an existing group and existing version for the group, therefore showing nothing in unhealthyResources at the end. Moving on to use full resource discovery instead - there is no way around.

Cleaned up imports. I have no convenient way of disabling the IDE features around structuring imports. The number of unrelated files with those restructured imports is so huge due to the fact that I had to manually switch import paths from `fusor` to `konveyor` and it caused autosaving and autoformatting. Most of those are now cleaned.

Removed autodowngrade from my previous PR this was based on.

gildub commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,414 @@
+	return resources, nil
+}
+
+func (r *Compare) unsupportedResources(resourceDiff []*metav1.APIResourceList) ([]schema.GroupVersionResource, error) {
```
There is no need here for a function type.  (r *Compare) is not used.

Danil-Grigorev replied:
Fixed, thanks :)

djwhatle commented on pkg/controller/migplan/gvk.go:
```diff
@@ -0,0 +1,414 @@
+}
+
+func reportGVK(plan *migapi.MigPlan, unsupportedMapping map[string][]schema.GroupVersionResource) {
+	unsupportedNamespaces := []migapi.UnsupportedNamespace{}
```
`incompatible` would be a more accurate term across the board in this PR. Please change all instances of `unsupported` to `incompatible`

djwhatle commented on pkg/apis/migration/v1alpha1/migplan_types.go:
```diff
@@ -57,7 +58,8 @@ type MigPlanSpec struct {
 type MigPlanStatus struct {
 	Conditions
-	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	UnsupportedNamespaces []UnsupportedNamespace `json:"unsupportedNamespaces,omitempty"`
```
Unsupported -> Incompatible

djwhatle commented on config/crds/migration_v1alpha1_migplan.yaml:
```diff
@@ -96,6 +96,22 @@ spec:
             observedGeneration:
               format: int64
               type: integer
+            unsupportedNamespaces:
```
Unsupported -> Incompatible

Danil-Grigorev commented:
@djwhatle Updated with requested changes included, with a minor note - we are dealing with GVRs but not GVKs, therefore the name change. @jortel Is it ok from your point of view?

jortel commented on config/crds/migration_v1alpha1_migplan.yaml:
```diff
@@ -96,6 +96,22 @@ spec:
+                type: object
+              type: array
+          required:
+          - observedGeneration
```
This change is reverting https://github.com/konveyor/mig-controller/pull/422.  The `observedGeneration` needs to be optional.

jortel commented on pkg/apis/migration/v1alpha1/migplan_types.go:
```diff
@@ -57,7 +58,8 @@ type MigPlanSpec struct {
 	Conditions
-	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	Incompatible
+	ObservedGeneration int64 `json:"observedGeneration"`
```
This change ^ is reverting https://github.com/konveyor/mig-controller/pull/422.  The `observedGeneration` needs to be `omitempty`.

Danil-Grigorev replied:
Oh, ok, I see. This resolves the issue, thanks @jortel

jortel commented on pkg/apis/migration/v1alpha1/gvr.go:
```diff
@@ -0,0 +1,32 @@
+// to contain resources incompatible by the migration
+type IncompatibleNamespace struct {
+	Name string            `json:"name"`
+	GVRs []IncompatibleGVR `json:"gvrs"`
```
Can you help me understand why this is GVR instead of GVK?  The field seems to contain kinds (Eg: pods).  Since resources are instances of kinds, I would expect a field named GVR to identify instances by name.

Danil-Grigorev replied:
For the most part it is explained by the dynamic client API, on which the feature was built on. It expects a `GVR` structure to fetch the resource, and the actual Kind was lost during transformations. It is not suitable, as the resources are lowercase and in plural tense, which allows it to be directly translated to the API path. You could fetch `apis/extensions/v1beta1/replicasets`, but server does not recognize `apis/extensions/v1beta1/ReplicaSet` as a valid path.

djwhatle commented on pkg/apis/migration/v1alpha1/gvr.go:
```diff
@@ -0,0 +1,32 @@
+type IncompatibleGVR struct {
+	Group    string `json:"group"`
+	Version  string `json:"version"`
+	Resource string `json:"resoruce"`
```
`json:"resoruce"` -> `json:"resource`

I see several places in this PR where "resource" is has typos in its spelling, @Danil-Grigorev let's fix these

Danil-Grigorev replied:
Fixed 😥

djwhatle commented:
Testing now.
